plugin/kubernetai/kubernetai.go

```diff
@@ -88,23 +90,43 @@ func (k8i Kubernetai) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns
 // AutoPath routes AutoPath requests to the authoritative kubernetes.
 func (k8i Kubernetai) AutoPath(state request.Request) []string {
 	var searchPath []string
+
+	// Abort if zone is not in kubernetai stanza.
+	var zMatch bool
+	for _, k8s := range k8i.Kubernetes {
+		zone := plugin.Zones(k8s.Zones).Matches(state.Name())
+		if zone != "" {
+			zMatch = true
+			break
+		}
+	}
+	if !zMatch {
+		return nil
+	}
+
+	// Add autopath result for the handled zones
 	for _, k := range k8i.Kubernetes {
-		zones := make([]string, 0, len(k.Zones)*2)
-		zones = append(zones, k.Zones...)
+		pod := k8i.p.PodWithIP(*k, state.IP())
+		if pod == nil {
+			return nil
+		}
+
+		search := make([]string, 3)
 		for _, z := range k.Zones {
-			if !strings.HasPrefix(z, "svc.") {
-				zones = append(zones, "svc."+z)
+			if z == "." {
+				search[0] = pod.Namespace + ".svc."
+				search[1] = "svc."
+				search[2] = "."
+			} else {
+				search[0] = pod.Namespace + ".svc." + z
+				search[1] = "svc." + z
+				search[2] = z
 			}
+			searchPath = append(search, searchPath...)
 		}
-		zone := plugin.Zones(zones).Matches(state.Name())
-		if zone != "" {
-			searchPath = append([]string{zone}, searchPath...)
-		}
-		searchPath = append(searchPath, zones...)
 	}
-
+	searchPath = append(searchPath, k8i.autoPathSearch...)
 	searchPath = append(searchPath, "")
```

Comment:
Do you still need this line given that `append` is being called on 128?

Reply:
yeah basically from what i read the autopath searchPath allways need to end with an empty string. 
It need to have an empty string for when you do a lookup for and exact match.
For example when the program ask for `svc-1-a.test-1.svc.cluster.local.` the ndot will add the search path transforming it to `svc-1-a.test-1.svc.cluster.local.test-1.svc.cluster.local.`
What the Servedns does in the module is getting the base path by striping part of the query 
Normally this `.test-1.svc.cluster.local.` and concatenating it with the result from the base path.
With empty string you get back on your feet and the request continue with `svc-1-a.test-1.svc.cluster.local.`

The issue with this is that it bring an unjust overhead on request that are properly scoped.
Unfortunately there is no way to know if the trailing `.test-1.svc.cluster.local.` comes from the resolv.conf or from the applications.